src/nvim/eval.c

```diff
@@ -7306,18 +7306,14 @@ static void f_changenr(typval_T *argvars, typval_T *rettv, FunPtr fptr)
  */
 static void f_char2nr(typval_T *argvars, typval_T *rettv, FunPtr fptr)
 {
-  if (has_mbyte) {
-    int utf8 = 0;
-
-    if (argvars[1].v_type != VAR_UNKNOWN) {
-      utf8 = tv_get_number_chk(&argvars[1], NULL);
+  if (argvars[1].v_type != VAR_UNKNOWN) {
+    if (!tv_check_num(&argvars[1])) {
+      return;
     }
-
-    rettv->vval.v_number = (utf8 ? *utf_ptr2char : *mb_ptr2char)(
-        (const char_u *)tv_get_string(&argvars[0]));
-  } else {
-    rettv->vval.v_number = (uint8_t)(tv_get_string(&argvars[0])[0]);
   }
+
+  rettv->vval.v_number = utf_ptr2char(
```

Comment:
In the current state I do not like the names: `mb_ptr2char` is using `utf_ptr2char` which yields one character. `mb_ptr2len` on the other side is using `utfc_ptr2len` which treats composing characters as a part of the character. Both start with `mb_`, but use different definitions of “character” (see the comment near `utfc_ptr2len`, it expcilitly names that cluster a character). I would more like keeping only `utf*` variants then only `mb_*`.